### PR TITLE
Add `allequal`

### LINKED ISSUE #43353: Add `allequal`
Since we already have `allunique`, it might make sense to also add `allequal`. One reason to add `allequal` is that a proper implementation is much faster than the naive `length(unique(itr)) == 1`.

```julia
function allequal(x)
    first, rest = Iterators.peel(x)

    for v in rest
        if v == first
            continue
        else
            return false
        end
    end

    true
end

slow_allequal(x) = length(unique(x)) == 1
```

```julia
julia> using BenchmarkTools

julia> x = randn(1000); y = fill(1, 1000);

julia> @btime allequal($x);
  3.288 ns (0 allocations: 0 bytes)

julia> @btime slow_allequal($x);
  36.791 μs (27 allocations: 65.95 KiB)

julia> @btime allequal($y);
  475.082 ns (0 allocations: 0 bytes)

julia> @btime slow_allequal($y);
  3.816 μs (6 allocations: 608 bytes)
```

### PR DESCRIPTION
**Pull request by CameronBieganek**
Fixes #43353.

### CODE REVIEW

**Comment by oscardssmith** [comment]
Adding to triage to decide if we want this.

**Comment by rfourquet** [comment]
Very much in favor, and this was just on my todo list. This is a feature which is requested from time to time and easy to get wrong. For example, on a zulip thread, the OP thought `allunique` was doing `isequal`, and then someone proposed `foldl(==, collection)` as an implementation for `allequal` (which works only with the most simple collections), which was also mentioned in a separate thread.

File: base/set.jl
Comment (jw3126) [inline base/set.jl:454]: I find this surprising. My expectation was `allequal(Dict(1=>1, 2=>1)) == true`.

Comment (rfourquet) [reply base/set.jl:454]: `allequal`'s behavior should probably not discriminate on container's type and work the same for all iterators, and `Dict` iterates over key-value pairs.

Comment (jw3126) [reply base/set.jl:454]: Could `allequal(Dict(1=>1, 2=>1))` be added to the doctest?

**Comment by timholy** [comment]
I'm opposed to growing our exports for essentially trivial functions.

File: base/set.jl
Comment (CameronBieganek) [reply base/set.jl:457]: Sure, I just added a similar example to the doctest.

**Comment by CameronBieganek** [comment]
I don't personally care whether `allequal` gets added to Base, but here are the two arguments that I see in favor:

1. `allequal` complements `allunique`, which is already in Base.
2. Users implementing this function themselves might be likely to define `allequal(itr) = length(unique(itr)) <= 1`, which is much less efficient.

Of course the "users might implement it inefficiently" argument could be used for lots of different functions, so it's probably not a very strong argument.

**Comment by CameronBieganek** [comment]
There are some errors in the CI tests. It seems like they're not related to the unit tests that I added, but I can't tell for sure.

**Comment by goretkin** [comment]
> I'm opposed to growing our exports for essentially trivial functions.

What about not exporting it, then?

For me, `allequal` is kind of in the same category as `only`. `only` is trivial. When you're feeling lazy or reckless, you can just use `first`. You can also just implement `allequal(x) = all(==(first(x), x)`, but it's even more wrong than using `first` in place of `only`.

**Comment by CameronBieganek** [comment]
> You can also just implement `allequal(x) = all(==(first(x), x)`, but it's even more wrong than using `first` in place of `only`.

Actually, there are scenarios where you might reasonably want `all(==(first(x), x)`. For example, if you want to use three-valued logic:

```julia
julia> x = [missing, missing];

julia> all(isequal(first(x)), x)
true

julia> all(==(first(x)), x)
missing
```

If you're looking to use three-valued logic, then you expect `allequal([missing, missing])` to be `missing`. One might also want `allequal([NaN, NaN])` to return `false`:

```julia
julia> y = [NaN, NaN];

julia> all(isequal(first(y)), y)
true

julia> all(==(first(y)), y)
false
```

I used `isequal` in this PR because `allunique` uses `isequal` (actually, the implementation does not, but that's the API that `allunique` presents). And of course `allunique` uses `isequal` because `unique` uses `isequal`, and `unique` uses `isequal` because sets and dictionaries use `isequal`.

**Comment by goretkin** [comment]
I should have have used `isequal`. I intended only to make a point about empty collections, which requires a `length` check, as does implementing `only`.

**Comment by Moelf** [comment]
slightly against, the justification for `allunique` (IMO) is the large speed up + non-trivial implementation:
```julia
julia> @benchmark allunique(x) setup=(x = rand(1:10, 100))
BenchmarkTools.Trial: 10000 samples with 962 evaluations.
 Range (min … max):   83.630 ns …   2.047 μs  ┊ GC (min … max):  0.00% … 93.80%
 Time  (median):     102.137 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):   118.439 ns ± 166.812 ns  ┊ GC (mean ± σ):  12.62% ±  8.44%

       ▁▁▁▄▃▅▆█▅▆▄▂▁                                             
  ▁▂▃▄▆█████████████▇▇▆▄▄▄▃▃▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  83.6 ns          Histogram: frequency by time          171 ns <

 Memory estimate: 400 bytes, allocs estimate: 4.

julia> @benchmark length(unique(x))==length(x) setup=(x = rand(1:10, 100))
BenchmarkTools.Trial: 10000 samples with 183 evaluations.
 Range (min … max):  558.109 ns …  11.964 μs  ┊ GC (min … max): 0.00% … 93.17%
 Time  (median):     592.109 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   632.985 ns ± 566.994 ns  ┊ GC (mean ± σ):  5.22% ±  5.48%

     ▁▄▃▅▅██▆▄▃▂                                                 
  ▁▂▄████████████▆▅▄▄▃▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  558 ns           Histogram: frequency by time          769 ns <

 Memory estimate: 880 bytes, allocs estimate: 7.
```

`allequal` seems to be something users can easily come up with & type

**Comment by mcabbott** [comment]
Have certainly written naive versions of both of these quite a few times. It is tidy to have shared verbs so you don't have to invent a name, or figure out someone else's -- like `only`, as mentioned.

Re timing `allunique`, at the size shown a different naive version is much faster -- perhaps Base's implementation should switch below some size?
```
julia> _allunique(x::AbstractVector) = all(!(x[i] in @view x[i+1:end]) for i in eachindex(x));

julia> @btime allunique(x) setup=(x = rand(1:10, 100));
  min 66.872 ns, mean 90.156 ns (4 allocations, 400 bytes)

julia> @btime _allunique(x) setup=(x = rand(1:10, 100));
  min 4.333 ns, mean 9.602 ns (0 allocations)
```

**Comment by JeffBezanson** [comment]
Triage is in favor. Personally I like simple utilities like this that save you a bit of cognitive load, provided the meaning is really obvious which I think it is in this case.

File: base/set.jl
Comment (Liozou) [inline base/set.jl:457]: I'm not sure whether it is right to add this specialization for `AbstractSet` and `AbstractDict`, since it could be wrong for multisets and multidicts (it is currently wrong with respect to the `Multiset` object of package `Multisets.jl`). All the more so because I can't think of any interest of the `allequal` functions on dicts and sets except for the multiset and multidict kinds! Also it can error if `length` is not defined, which is not a requirement for `AbstractSet` and `AbstractDict` AFAIK.
If you agree, it can still be kept for the relevant Base cases of course (`Set`, `IdSet`, `KeySet`, `BitSet`; `Dict`, `IdDict`, `Pairs`, `WeakKeyDict`, `ImmutableDict`, `EnvDict`).

Comment (CameronBieganek) [reply base/set.jl:457]: I understand your point, but I'm not sure whether it makes sense for a `Multiset` to be a subtype of `AbstractSet`. In fact, `Multiset` in Multisets.jl was not a subtype of `AbstractSet` until the recent 0.4.0 version (the current version is 0.4.2).

I just modeled the `allequal` method for `AbstractSet` and `AbstractDict` after the `allunique` method that already exists:

```julia
allunique(::Union{AbstractSet,AbstractDict}) = true
```

As you can see, the existing `allunique` method is incorrect for multisets. So I'm not sure what the best course of action is.

Comment (Liozou) [reply base/set.jl:457]: Oh that's a very good point. Being consistent is probably more important than just one edge-case, so I will try to remember submitting a PR to `Multisets` once this PR is merged then, that should take care of the issue. Thanks!

Comment (rfourquet) [reply base/set.jl:457]: I wish we had in `Base` a supertype of `AbstractDict` to include multisets, say `Associative`, and similarly for sets. But I otherwise agree that `AbstractSet` generally assumes uniqueness of keys.

**Comment by CameronBieganek** [comment]
I added a NEWS.md entry and updated the docstrings, including a compat entry for the `allequal` docstring. So, I think this is ready for a final review and merge.

I'm not sure where CI went... I don't see any CI results here anymore.

**Review by rfourquet — APPROVED** [review]

**Comment by CameronBieganek** [comment]
@oscardssmith Can this be merged before the 1.8 deadline?

**Comment by oscardssmith** [comment]
LGTM

**Comment by CameronBieganek** [comment]
If anyone is listening, this should be good to merge before the 1.8 deadline tomorrow. (I don't have merge rights.)

**Pull request merged**
